op/model.py

```diff
@@ -231,26 +231,28 @@ def _load(self):
         return self._backend.config_get()
 
 
-class ResourceMapping(Mapping):
-    """Map of resource names to the local path for the resource."""
-    def __init__(self, resource_names, backend):
+class Resource:
+    """Object representing a resource for the charm.
+    """
+    def __init__(self, name, backend):
+        self.name = name
         self._backend = backend
-        self._data = {resource_name: None for resource_name in resource_names}
+        self.path = None
 
-    def __contains__(self, key):
-        return key in self._data
-
-    def __len__(self):
-        return len(self._data)
-
-    def __iter__(self):
-        return iter(self._data)
+    def fetch(self):
+        """Fetch the resource from the controller or store.
 
-    def __getitem__(self, resource_name):
-        resource_path = self._data[resource_name]
-        if resource_path is None:
-            resource_path = self._data[resource_name] = Path(self._backend.resource_get(resource_name))
-        return resource_path
+        If successfully fetched, this returns True and the resource file can be
+        accessed via the path attribute.
+        """
+        try:
+            filename = self._backend.resource_get(self.name)
+            if filename:
+                self.path = Path(filename)
+                return True
+        except CalledProcessError:
```

Comment:
We need to polish that error hierarchy before 1.0 is out.

Reply:
AFAICT, `resource-get` only ever either returns a filename or exits with a return code of 1 if it was unable to fetch the resource for any reason, so I'm not really sure we can improve this case any. But I'm open to suggestions or corrections if I'm mistaken.